Don't normalize in AstConv

aliemjay opened the pull request:
See individual commits.

Fixes #101350
Fixes #54940

aliemjay committed:
don't normalize in astconv

We delay projection normalization to further stages in order to
register user type annotations before normalization in HIR typeck.

There are two consumers of astconv: ItemCtxt and FnCtxt.
The former already expects unnormalized types from astconv, see its
AstConv trait impl.
The latter needs `RawTy` for a cleaner interface.

Unfortunately astconv still needs the normalization machinery in
order to resolve enum variants that have projections in the self type,
e.g. `<<T as Trait>::Assoc>::StructVariant {}`.
This is why `AstConv::normalize_ty_2` is necessary.

rust-highfive commented:
r? @jackh726

(rust-highfive has picked a reviewer for you, use r? to override)

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3329820941/jobs/5507522299) [(plain)](https://github.com/rust-lang/rust/commit/c3bf78b4be89a676a679a4561004c7a80fce43e6/checks/9119219779/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
* highest error code: E0790
Found 506 error codes
Found 0 error(s) in error codes
Done!
tidy error: /checkout/compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs:1214: TODO is deprecated; use FIXME
tidy error: /checkout/compiler/rustc_hir_typeck/src/closure.rs:607: TODO is deprecated; use FIXME
tidy error: /checkout/src/test/ui/tests2/tto1.rs:58: TODO is deprecated; use FIXME
tidy error: /checkout/compiler/rustc_hir_typeck/src/method/confirm.rs:413: TODO is deprecated; use FIXME
some tidy checks failed
Build completed unsuccessfully in 0:00:11
```

</details>

the8472 commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 82fa1136489005893be02555b865ca6a90d923f3 with merge 1ab52f9b015fe3b11f0862dd6cbf44c426995a28...
<!-- homu: {"type":"TryBuildStarted","head_sha":"82fa1136489005893be02555b865ca6a90d923f3","merge_sha":"1ab52f9b015fe3b11f0862dd6cbf44c426995a28"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3335664581/jobs/5521804951)
Build commit: 1ab52f9b015fe3b11f0862dd6cbf44c426995a28 (`1ab52f9b015fe3b11f0862dd6cbf44c426995a28`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3335664581/jobs/5521804951"},"merge_sha":"1ab52f9b015fe3b11f0862dd6cbf44c426995a28"} -->

rust-timer commented:
Queued 1ab52f9b015fe3b11f0862dd6cbf44c426995a28 with parent 44fcfb0a9606fa08ea293a71f2e6d470227ef9df, future [comparison URL](https://perf.rust-lang.org/compare.html?start=44fcfb0a9606fa08ea293a71f2e6d470227ef9df&end=1ab52f9b015fe3b11f0862dd6cbf44c426995a28).

rust-timer commented:
Finished benchmarking commit (1ab52f9b015fe3b11f0862dd6cbf44c426995a28): [comparison URL](https://perf.rust-lang.org/compare.html?start=44fcfb0a9606fa08ea293a71f2e6d470227ef9df&end=1ab52f9b015fe3b11f0862dd6cbf44c426995a28&stat=instructions:u).

### Overall result: no relevant changes - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=44fcfb0a9606fa08ea293a71f2e6d470227ef9df&end=1ab52f9b015fe3b11f0862dd6cbf44c426995a28&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=44fcfb0a9606fa08ea293a71f2e6d470227ef9df&end=1ab52f9b015fe3b11f0862dd6cbf44c426995a28&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=44fcfb0a9606fa08ea293a71f2e6d470227ef9df&end=1ab52f9b015fe3b11f0862dd6cbf44c426995a28&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

aliemjay committed:
introduce AstConv::probe_adt

aliemjay committed:
fix method substs

aliemjay committed:
fix struct path

aliemjay commented:
r? @lcnr

bors commented:
:umbrella: The latest upstream changes (presumably #102618) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

lcnr commented on compiler/rustc_hir_typeck/src/fn_ctxt/mod.rs:
```diff
@@ -279,19 +278,39 @@ impl<'a, 'tcx> AstConv<'tcx> for FnCtxt<'a, 'tcx> {
+    fn probe_adt(&self, span: Span, ty: Ty<'tcx>) -> Option<ty::AdtDef<'tcx>> {
+        match ty.kind() {
+            ty::Adt(adt_def, _) => Some(*adt_def),
+            // FIXME: We should normalize even with escaping bound vars.
```
no, we shouldn't imo

imo we should have never relied on normalization during ast lowering. We won't be able to remove the existing normalization due to backwards compatibility. I think we should just live with the existing inconsistency, potentially even linting when normalization is required to remove it 2030 or whatever

aliemjay replied:
I tend to agree. The fact that path resolution depends on normalization may be an unintended behavior.

lcnr replied:
cc @rust-lang/types I generally don't think we should normalize during astconv to resolve stuff like `Self::A` if `Self` is a projection. Examples include
```rust
trait Identity {
    type Id;
}
impl<T> Identity for T {
    type Id = T;
}

enum Foo {
    A,
    B,
}

trait Trait {
    fn return_a() -> Self;
}

impl Trait for <Foo as Identity>::Id {
    fn return_a() -> Self {
        Self::A
    }
}

fn main() {
    let x = <<Foo as Identity>::Id>::A;
}
```
We do have to keep supporting this for backwards compatibility however.

I therefore think that we should 1) remove the FIXME here and add a comment for "we cannot normalize inside of a binder" 2) rename this method to `probe_adt_normalize_for_backcompat` to prevent us from at least adding further uses of that method.

aliemjay replied:
This looks like T-lang issue. Can we open a new issue and nominate it for t-lang?

Anyway  I toned down the FIXME comment and opened a new ticket for it #104767. See also #103640.

lcnr commented on src/test/ui/associated-types/bound-region-variant-resolution.rs:
```diff
@@ -0,0 +1,15 @@
+// This is a bug!
```
can you open an issue for that? You can then add `// known-bug: #<issuenum>` to this file

lcnr commented:
Thanks :heart: 

As this is a breaking change, going to run crater and then go through a t-types FCP. r=me on the impl after some nits

@bors try

bors commented:
:hourglass: Trying commit a1a29a30952de6b1aed11075b0979825595c5275 with merge c4ca63acfd7111dc43e5bc36534474f323ca449a...
<!-- homu: {"type":"TryBuildStarted","head_sha":"a1a29a30952de6b1aed11075b0979825595c5275","merge_sha":"c4ca63acfd7111dc43e5bc36534474f323ca449a"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3409453092/jobs/5673595731)
Build commit: c4ca63acfd7111dc43e5bc36534474f323ca449a (`c4ca63acfd7111dc43e5bc36534474f323ca449a`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3409453092/jobs/5673595731"},"merge_sha":"c4ca63acfd7111dc43e5bc36534474f323ca449a"} -->

aliemjay committed:
make ascribe_user_type a TypeOp

Projection types in user annotations may contain inference variables.
This makes the normalization depend on the unification with the actual
type and thus requires a separate TypeOp to track the obligations.
Otherwise simply calling `TypeChecker::normalize` would ICE with
"unexpected ambiguity"

bors commented:
:umbrella: The latest upstream changes (presumably #104481) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3478440286/jobs/5815810858) [(plain)](https://github.com/rust-lang/rust/commit/0b20890b5bc6bb9fbca26a8bfc2d3733a783c313/checks/9521907123/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
........................................................................................ 8448/13852
........................................................................................ 8536/13852
........................................................................................ 8624/13852
........................................................................................ 8712/13852
......F...............................F..................i..ii.......................... 8800/13852
...............................................iiii..................................... 8976/13852
........................................................................................ 9064/13852
i........................................i.............................................. 9152/13852
......................i................................................................. 9240/13852
---
---- [ui] src/test/ui/nll/user-annotations/ascribed-type-wf.rs stdout ----

error: test compilation failed although it shouldn't!
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/nll/user-annotations/ascribed-type-wf.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/nll/user-annotations/ascribed-type-wf" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/nll/user-annotations/ascribed-type-wf/auxiliary"
stdout: none
--- stderr -------------------------------
error: lifetime may not live long enough
   |
   |
LL | fn extend<'a>() {
   |           -- lifetime `'a` defined here
LL |     None::<<&'a () as Trait>::Ty>;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ requires that `'a` must outlive `'static`
error: aborting due to previous error
------------------------------------------



---- [ui] src/test/ui/nll/user-annotations/normalization-infer.rs stdout ----
diff of stderr:

Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
79   --> $DIR/normalization-infer.rs:32:28
80    |
81 LL |     let _: Alias<_, _> = (&temp(), 0u8);
-    |            -----------     ^^^^^^ creates a temporary which is freed while still in use
+    |            -----------     ^^^^^^ creates a temporary value which is freed while still in use
83    |            |
84    |            type annotation requires that borrow lasts for `'static`


92 LL |     Some::<Alias<_, _>>((&temp(), 0u8));
93    |                         --^^^^^^------ - temporary value is freed at the end of this statement
-    |                         | creates a temporary which is freed while still in use
+    |                         | creates a temporary value which is freed while still in use
+    |                         | creates a temporary value which is freed while still in use
96    |                         this usage requires that borrow lasts for `'static`
98 error: aborting due to 9 previous errors


The actual stderr differed from the expected stderr.
The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/nll/user-annotations/normalization-infer/normalization-infer.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args nll/user-annotations/normalization-infer.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/nll/user-annotations/normalization-infer.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/nll/user-annotations/normalization-infer" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/nll/user-annotations/normalization-infer/auxiliary"
stdout: none
--- stderr -------------------------------
error[E0310]: the parameter type `A` may not live long enough
   |
   |
LL |     let _: <(_,) as Tr>::Ty = a; //~ ERROR type `A`
   |            ^^^^^^^^^^^^^^^^ ...so that the type `A` will meet its required lifetime bounds
help: consider adding an explicit lifetime bound...
   |
   |
LL | fn test1<A: 'static, B, C, D>(a: A, b: B, c: C) {

error[E0310]: the parameter type `B` may not live long enough
  --> /checkout/src/test/ui/nll/user-annotations/normalization-infer.rs:12:5
   |
   |
LL |     Some::<<(_,) as Tr>::Ty>(b); //~ ERROR type `B`
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ ...so that the type `B` will meet its required lifetime bounds
help: consider adding an explicit lifetime bound...
   |
   |
LL | fn test1<A, B: 'static, C, D>(a: A, b: B, c: C) {

error[E0310]: the parameter type `C` may not live long enough
  --> /checkout/src/test/ui/nll/user-annotations/normalization-infer.rs:13:11
   |
   |
LL |     || -> <(_,) as Tr>::Ty { c }; //~ ERROR type `C`
   |           ^^^^^^^^^^^^^^^^ ...so that the type `C` will meet its required lifetime bounds
help: consider adding an explicit lifetime bound...
   |
   |
LL | fn test1<A, B, C: 'static, D>(a: A, b: B, c: C) {

error[E0310]: the parameter type `D` may not live long enough
  --> /checkout/src/test/ui/nll/user-annotations/normalization-infer.rs:14:6
   |
   |
LL |     |d: <(_,) as Tr>::Ty| -> D { d }; //~ ERROR type `D`
   |      ^ ...so that the type `D` will meet its required lifetime bounds
help: consider adding an explicit lifetime bound...
   |
   |
LL | fn test1<A, B, C, D: 'static>(a: A, b: B, c: C) {

error[E0310]: the parameter type `A` may not live long enough
  --> /checkout/src/test/ui/nll/user-annotations/normalization-infer.rs:28:12
   |
   |
LL |     let _: Alias<_, _> = (a, 0u8); //~ ERROR type `A`
   |            ^^^^^^^^^^^ ...so that the type `A` will meet its required lifetime bounds
help: consider adding an explicit lifetime bound...
   |
   |
LL | fn test2<A: 'static, B, C>(a: A, b: B, c: C) {

error[E0310]: the parameter type `B` may not live long enough
  --> /checkout/src/test/ui/nll/user-annotations/normalization-infer.rs:29:5
   |
   |
LL |     Some::<Alias<_, _>>((b, 0u8)); //~ ERROR type `B`
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ...so that the type `B` will meet its required lifetime bounds
help: consider adding an explicit lifetime bound...
   |
   |
LL | fn test2<A, B: 'static, C>(a: A, b: B, c: C) {

error[E0310]: the parameter type `C` may not live long enough
  --> /checkout/src/test/ui/nll/user-annotations/normalization-infer.rs:30:11
   |
   |
LL |     || -> Alias<_, _> { (c, 0u8) }; //~ ERROR type `C`
   |           ^^^^^^^^^^^ ...so that the type `C` will meet its required lifetime bounds
help: consider adding an explicit lifetime bound...
   |
   |
LL | fn test2<A, B, C: 'static>(a: A, b: B, c: C) {

error[E0716]: temporary value dropped while borrowed
  --> /checkout/src/test/ui/nll/user-annotations/normalization-infer.rs:32:28
   |
   |
LL |     let _: Alias<_, _> = (&temp(), 0u8); //~ ERROR temporary value
   |            -----------     ^^^^^^ creates a temporary value which is freed while still in use
   |            |
   |            type annotation requires that borrow lasts for `'static`
LL | }
   | - temporary value is freed at the end of this statement

error[E0716]: temporary value dropped while borrowed
error[E0716]: temporary value dropped while borrowed
  --> /checkout/src/test/ui/nll/user-annotations/normalization-infer.rs:33:27
   |
LL |     Some::<Alias<_, _>>((&temp(), 0u8)); //~ ERROR temporary value
   |                         --^^^^^^------ - temporary value is freed at the end of this statement
   |                         | creates a temporary value which is freed while still in use
   |                         | creates a temporary value which is freed while still in use
   |                         this usage requires that borrow lasts for `'static`
error: aborting due to 9 previous errors

Some errors have detailed explanations: E0310, E0716.
For more information about an error, try `rustc --explain E0310`.
```

</details>

aliemjay committed:
remove unnecessary normalize call

aliemjay commented on compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs:
```diff
@@ -1187,7 +1161,8 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         let (res, self_ctor_substs) = if let Res::SelfCtor(impl_def_id) = res {
-            let ty = self.normalize_ty(span, tcx.at(span).type_of(impl_def_id));
+            let ty = tcx.at(span).type_of(impl_def_id);
+            let ty = self.normalize_associated_types_in(span, ty);
```
`SelfCtor` type is already checked by item `wfcheck`, so I don't think we will miss region obligations when normalizing.

aliemjay replied:
This was indeed problematic. I have fixed it in https://github.com/rust-lang/rust/pull/101947/commits/e05158d6c8b0ca62888140195a93a3487572b028.

aliemjay commented:
@rustbot review

jyn514 commented:
@bors try

bors commented:
:hourglass: Trying commit f422dc15e018db04bea06a20379d997e365a8151 with merge fbef200192438a84822be119ec8070ded9a19487...
<!-- homu: {"type":"TryBuildStarted","head_sha":"f422dc15e018db04bea06a20379d997e365a8151","merge_sha":"fbef200192438a84822be119ec8070ded9a19487"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3485463477/jobs/5832160094)
Build commit: fbef200192438a84822be119ec8070ded9a19487 (`fbef200192438a84822be119ec8070ded9a19487`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3485463477/jobs/5832160094"},"merge_sha":"fbef200192438a84822be119ec8070ded9a19487"} -->

Dylan-DPC commented:
@craterbot check crates=list:https://crater-reports.s3.amazonaws.com/pr-101947/retry-regressed-list.txt

Dylan-DPC commented:
@craterbot check

lcnr commented:
I think https://crates.io/crates/abstract-add-on/0.2.0-beta.2 might be the only actual regression? :thinking: I can't find the source for that crate. Considering that this version has less than 50 downloads this breakage feels acceptable. cc @CyberHoward

Would be good for someone else to also have a look at the crater results as only having 1 regression is actually pretty surprising to me, would have expected a bigger impact.

CyberHoward commented:
Thanks for the cc. Could you share any resources on why my code shouldn't have compiled in the first place?

aliemjay commented:
> Would be good for someone else to also have a look at the crater results as only having 1 regression is actually pretty surprising to me, would have expected a bigger impact.

@lcnr I think the impact is not as big as you expect because we're still not doing WF-checks on  user annotations prior to normalization, so this code is still accepted:
```rust
trait Trait { type Ty; }
impl<T> Trait for T { type Ty = (); }
fn test<T>() {
    let _: <&'static T as Trait>::Ty = ();
}
```

I consider this a different issue than #101350 because it is a more long-standing problem and not strictly an NLL regression.

aliemjay commented:
> Thanks for the cc. Could you share any resources on why my code shouldn't have compiled in the first place?

@CyberHoward you can take a look at the error message: https://crater-reports.s3.amazonaws.com/pr-101947-1/try%23fbef200192438a84822be119ec8070ded9a19487/reg/abstract-add-on-0.2.0-beta.2/log.txt

In `impl ExecuteEndpoint for AddonContract<E>` you have a bound `E: 'static` but it's not present at the error site.

aliemjay committed:
rename create_raw_ty -> handle_raw_ty

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3530451277/jobs/5922458294) [(plain)](https://github.com/rust-lang/rust/commit/c5d1da178b1d612cc4239b346635bdd361b8e129/checks/9660320310/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
1 error: expected type, found variant `Some`
-   --> $DIR/bound-region-variant-resolution.rs:9:23
+   --> $DIR/bound-region-variant-resolution.rs:10:23
3    |
4 LL |     None::<for<'y> fn(<&'x Option<()> as Deref>::Target::Some)>;

6 
7 error[E0223]: ambiguous associated type
-   --> $DIR/bound-region-variant-resolution.rs:11:23
-   --> $DIR/bound-region-variant-resolution.rs:11:23
+   --> $DIR/bound-region-variant-resolution.rs:12:23
9    |
10 LL |     None::<for<'y> fn(<&'y Option<()> as Deref>::Target::Some)>;
11    |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use fully-qualified syntax: `<<&'y Option<()> as Deref>::Target as Trait>::Some`

The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/associated-types/bound-region-variant-resolution/bound-region-variant-resolution.stderr
To update references, rerun the tests and pass the `--bless` flag
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args associated-types/bound-region-variant-resolution.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/associated-types/bound-region-variant-resolution.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/associated-types/bound-region-variant-resolution" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/associated-types/bound-region-variant-resolution/auxiliary"
stdout: none
--- stderr -------------------------------
error: expected type, found variant `Some`
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
   |
   |
LL |     None::<for<'y> fn(<&'x Option<()> as Deref>::Target::Some)>;

error[E0223]: ambiguous associated type
  --> /checkout/src/test/ui/associated-types/bound-region-variant-resolution.rs:12:23
   |
   |
LL |     None::<for<'y> fn(<&'y Option<()> as Deref>::Target::Some)>;
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use fully-qualified syntax: `<<&'y Option<()> as Deref>::Target as Trait>::Some`
error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0223`.
------------------------------------------
```

</details>

lcnr commented:
While this PR unblocks future work, it's only user-facing impact is that we now check the constraints used **to** normalize projections in user-types during borrowck. We do not yet check that the types used in projections themselves are well-formed.

That is implemented #104746 which needs a separate crater run. We found a single regression on crater as stated in https://github.com/rust-lang/rust/pull/101947#issuecomment-1321915979.

```rust
trait Trait {
    type Ty;
}

impl Trait for &'static () {
    type Ty = ();
}
impl<'a, 'b> Trait for &'a &'b () {
    type Ty = ();
}

fn ok_to_err<'a>() {
    None::<<&'a () as Trait>::Ty>;
    //~^ ERROR lifetime may not live long enough
}

fn still_accepted<'a>() {
    None::<<&'a &'static () as Trait>::Ty>;
    // The `'a: 'static` bound is part of the WF requirements
    // of `&'a &'static ()`, not a bound on the impl.
}
```

@rfcbot fcp merge

bors commented:
:umbrella: The latest upstream changes (presumably #104935) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

aliemjay commented on compiler/rustc_hir_typeck/src/fn_ctxt/mod.rs:
```diff
@@ -282,19 +281,38 @@ impl<'a, 'tcx> AstConv<'tcx> for FnCtxt<'a, 'tcx> {
+/// This is a bridge between the interface of `AstConv`, which outputs a raw `Ty`,
+/// and the API in this module, which expect `Ty` to be fully normalized.
+#[derive(Clone, Copy, Debug)]
+pub struct RawTy<'tcx> {
+    pub raw: Ty<'tcx>,
+
+    /// The normalized form of `raw`, stored here for efficiency.
+    pub normalized: Ty<'tcx>,
+}
```
```suggestion
struct RawTy<'tcx> {
    raw: Ty<'tcx>,

    /// The normalized form of `raw`, stored here for efficiency.
    normalized: Ty<'tcx>,
}
```

aliemjay commented on src/test/ui/nll/user-annotations/normalization-2.rs:
```diff
@@ -0,0 +1,114 @@
+    fn method2<X>(&self) {}
+}
+
+type Ty<'a> = <&'a () as Trait>::Assoc;
```
add a similar test for inherent associated types. cc #105315.
```suggestion
type Ty<'a> = CustomTy::<'a>::Assoc;
```

aliemjay commented:
The FCP has been stale for a while, going to nominate it for further discussion
@rustbot label I-types-nominated

aliemjay commented:
needs a rebase and some minor tweaks
@rustbot author

Dylan-DPC commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 05b8055707e143ac2ff862438b1a09f86e641101 with merge abb2387f077f4fc60458a49b60eefae2def2b46d...
<!-- homu: {"type":"TryBuildStarted","head_sha":"05b8055707e143ac2ff862438b1a09f86e641101","merge_sha":"abb2387f077f4fc60458a49b60eefae2def2b46d"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3766273591/jobs/6404327465)
Build commit: abb2387f077f4fc60458a49b60eefae2def2b46d (`abb2387f077f4fc60458a49b60eefae2def2b46d`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3766273591/jobs/6404327465"},"merge_sha":"abb2387f077f4fc60458a49b60eefae2def2b46d"} -->

aliemjay commented:
@rust-timer queue

rust-timer commented:
Insufficient permissions to issue commands to rust-timer.
<!-- rust-timer: temporary -->

lqd commented:
@rust-timer build abb2387f077f4fc60458a49b60eefae2def2b46d

rust-timer commented:
Queued abb2387f077f4fc60458a49b60eefae2def2b46d with parent 14e2fe4039f0b42c1ac519671e27943560530458, future [comparison URL](https://perf.rust-lang.org/compare.html?start=14e2fe4039f0b42c1ac519671e27943560530458&end=abb2387f077f4fc60458a49b60eefae2def2b46d).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (abb2387f077f4fc60458a49b60eefae2def2b46d): [comparison URL](https://perf.rust-lang.org/compare.html?start=14e2fe4039f0b42c1ac519671e27943560530458&end=abb2387f077f4fc60458a49b60eefae2def2b46d&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=14e2fe4039f0b42c1ac519671e27943560530458&end=abb2387f077f4fc60458a49b60eefae2def2b46d&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 0.2% | [0.2%, 0.2%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=14e2fe4039f0b42c1ac519671e27943560530458&end=abb2387f077f4fc60458a49b60eefae2def2b46d&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 4.7% | [4.7%, 4.7%] | 1     |
| Regressions ❌ <br /> (secondary)  | 2.7% | [2.7%, 2.8%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 4.7% | [4.7%, 4.7%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=14e2fe4039f0b42c1ac519671e27943560530458&end=abb2387f077f4fc60458a49b60eefae2def2b46d&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

aliemjay committed:
more tests

aliemjay committed:
don't eagerly normalize SelfCtor type

Delay until user annotations are registered.
See the added test.

aliemjay commented:
@rustbot review
r? @lcnr

bors commented:
:umbrella: The latest upstream changes (presumably #106143) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

lcnr commented:
@bors r+ rollup=never

bors commented:
:pushpin: Commit bf228ace5cf6824078d6d36144ad8a65f07fa8d3 has been approved by `lcnr`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=lcnr bf228ace5cf6824078d6d36144ad8a65f07fa8d3 -->
<!-- homu: {"type":"Approved","sha":"bf228ace5cf6824078d6d36144ad8a65f07fa8d3","approver":"lcnr","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit bf228ace5cf6824078d6d36144ad8a65f07fa8d3 with merge af58fc869910eae0ec525d49cbb6d953ea10a8c6...
<!-- homu: {"type":"BuildStarted","head_sha":"bf228ace5cf6824078d6d36144ad8a65f07fa8d3","merge_sha":"af58fc869910eae0ec525d49cbb6d953ea10a8c6"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3875248220/jobs/6610554890)
Approved by: lcnr
Pushing af58fc869910eae0ec525d49cbb6d953ea10a8c6 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"lcnr","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3875248220/jobs/6610554890"},"merge_sha":"af58fc869910eae0ec525d49cbb6d953ea10a8c6"} -->

rust-timer commented:
Finished benchmarking commit (af58fc869910eae0ec525d49cbb6d953ea10a8c6): [comparison URL](https://perf.rust-lang.org/compare.html?start=89e0576bd3aec2f1acf935b353900bd75742860b&end=af58fc869910eae0ec525d49cbb6d953ea10a8c6&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=89e0576bd3aec2f1acf935b353900bd75742860b&end=af58fc869910eae0ec525d49cbb6d953ea10a8c6&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.1%  | [0.1%, 0.1%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 0.2%  | [0.2%, 0.2%]   | 3     |
| Improvements ✅ <br /> (primary)   | -1.0% | [-1.0%, -1.0%] | 1     |
| Improvements ✅ <br /> (secondary) | -0.8% | [-0.9%, -0.8%] | 7     |
| All ❌✅ (primary)                 | -0.5% | [-1.0%, 0.1%]  | 2     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=89e0576bd3aec2f1acf935b353900bd75742860b&end=af58fc869910eae0ec525d49cbb6d953ea10a8c6&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=89e0576bd3aec2f1acf935b353900bd75742860b&end=af58fc869910eae0ec525d49cbb6d953ea10a8c6&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.1% | [-1.1%, -1.1%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -1.1% | [-1.1%, -1.1%] | 1     |
</details>
